cloud/openstack/os_server.py

```diff
@@ -0,0 +1,440 @@
+      region_name: region-b.geo-1
+      availability_zone: az2
+      image: 9302692b-b787-4b52-a3a6-daebb79cb498
+      key_name: test
+      timeout: 200
+      flavor: 101
+      security_groups: default
+      auto_floating_ip: yes
+
+# Creates a new instance in named cloud mordred availability zone az2
+# and assigns a pre-known floating IP
+- name: launch a compute instance
+  hosts: localhost
+  tasks:
+  - name: launch an instance
+    os_server:
+      state: present
+      cloud: mordred
+      name: vm1
+      availability_zone: az2
+      image: 9302692b-b787-4b52-a3a6-daebb79cb498
+      key_name: test
+      timeout: 200
+      flavor: 101
+      floating-ips:
+        - 12.34.56.79
+
+# Creates a new instance with 4G of RAM on Ubuntu Trusty, ignoring
+# deprecated images
+- name: launch a compute instance
+  hosts: localhost
+  tasks:
+  - name: launch an instance
+    os_server:
+      name: vm1
+      state: present
+      cloud: mordred
+      region_name: region-b.geo-1
+      image: Ubuntu Server 14.04
+      image_exclude: deprecated
+      flavor_ram: 4096
+
+# Creates a new instance with 4G of RAM on Ubuntu Trusty on a Performance node
+- name: launch a compute instance
+  hosts: localhost
+  tasks:
+  - name: launch an instance
+    os_server:
+      name: vm1
+      cloud: rax-dfw
+      state: present
+      image: Ubuntu 14.04 LTS (Trusty Tahr) (PVHVM)
+      flavor_ram: 4096
+      flavor_include: Performance
+'''
+
+
+def _exit_hostvars(module, cloud, server, changed=True):
+    hostvars = meta.get_hostvars_from_server(cloud, server)
+    module.exit_json(changed=changed, id=server.id, openstack=hostvars)
+
+
+def _network_args(module, cloud):
+    args = []
+    for net in module.params['nics']:
+        if net.get('net-id'):
+            args.append(net)
+        elif net.get('net-name'):
+            by_name = cloud.get_network(net['net-name'])
+            if not by_name:
+                module.fail_json(
+                    msg='Could not find network by net-name: %s' %
+                    net['net-name'])
+            args.append({'net-id': by_name['id']})
+    return args
+
+
+def _delete_server(module, cloud):
+    try:
+        cloud.delete_server(
+            module.params['name'], wait=module.params['wait'],
+            timeout=module.params['timeout'])
+    except Exception as e:
+        module.fail_json(msg="Error in deleting vm: %s" % e.message)
+    module.exit_json(changed=True, result='deleted')
+
+
+def _create_server(module, cloud):
+    flavor = module.params['flavor']
+    flavor_ram = module.params['flavor_ram']
+    flavor_include = module.params['flavor_include']
+
+    image_id = cloud.get_image_id(
+        module.params['image'], module.params['image_exclude'])
+
+    if flavor:
+        flavor_id = cloud.get_flavor(flavor)
+    else:
+        flavor_id = cloud.get_flavor_by_ram(flavor_ram, flavor_include)
+
+    nics = _network_args(module, cloud)
+
+    bootkwargs = dict(
```

Comment:
Is it possible to add "block_device_mapping_v2" option to bootkwargs (probably should be mutually exclusive with root_volume)? It is a new syntax for attaching block devices introduced in the Nova v2 API, old syntax will be removed as of v3. Details here: https://wiki.openstack.org/wiki/BlockDeviceConfig